src/util.rs

```diff
@@ -0,0 +1,25 @@
+#[derive(Debug)]
+/// Allows us to wrap and return `impl Iterator` trait objects
+/// with different types, namely `iter::empty` and other iterators.
+pub enum Either<A, B> {
+    A(A),
+    B(B),
+}
```

Comment:
See comment in pass about the `Either` name.

Reply:
What do we get here that we don't get with a tuple? The wrapper?

Reply:
In short, the ability to return a (possibly empty) `impl Iterator<Item=T>` instead of an `Option<impl Iterator<Item=T>>`.

Take this function, for instance:

```rust
pub fn exports(&self) -> impl Iterator<Item = &ExportEntry> {
    self.module
        .export_section()
        .map_or(Either::Left(iter::empty()),
                |sec| Either::Right(sec.entries().iter()))
}
```

Here, we're returning a type that `impl Iterator<Item = &ExportEntry>`.

- If the module doesn't have an exports section, we return an `iter::empty()`.
- If the module *does* have an exports section, we return a `slice::Iter<'_, &ExportEntry>`.

But if we were to remove the wrapping calls to `Either::Left` and `Either::Right`, we'd get a compile error about the two branches of the `.map_or` returning different types, even though both of those types implement `Iterator`.

Reply:
Sounds good.